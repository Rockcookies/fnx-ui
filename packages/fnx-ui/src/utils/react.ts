import { Children, ReactElement } from 'react';
import { isFragment } from 'react-is';

export const toElementArray = (
	children: React.ReactNode,
	option?: { keepEmpty?: boolean },
): ReactElement[] => {
	const keepEmpty = option && option.keepEmpty;
	let ret: React.ReactElement[] = [];

	Children.forEach(children, (child: any) => {
		if ((child === undefined || child === null) && !keepEmpty) {
			return;
		}

		if (Array.isArray(child)) {
			ret = ret.concat(toElementArray(child));
		} else if (isFragment(child) && child.props) {
			ret = ret.concat(toElementArray(child.props.children, option));
		} else {
			ret.push(child);
		}
	});

	return ret;
};
