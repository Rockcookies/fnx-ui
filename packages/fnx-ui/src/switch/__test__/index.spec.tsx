import { fireEvent, render } from '@testing-library/react';
import React from 'react';
import Switch from '../index';

describe('<Switch/>', () => {
	it('should emit onChange event when click the switch button', () => {
		const onClick = jest.fn();
		const onChange = jest.fn();

		const { getByTestId } = render(
			<Switch
				data-testid="switch"
				onClick={onClick}
				onChange={onChange}
			/>,
		);

		fireEvent.click(getByTestId('switch'));

		expect(onChange).toBeCalledWith(true);

		fireEvent.click(getByTestId('switch'));

		expect(onChange).toBeCalledWith(false);
		expect(onClick).toBeCalledTimes(2);
	});

	it('should not emit event when disabled', () => {
		const onClick = jest.fn();
		const onChange = jest.fn();

		const { getByTestId } = render(
			<Switch
				data-testid="switch"
				disabled
				defaultChecked
				onClick={onClick}
				onChange={onChange}
			/>,
		);

		fireEvent.click(getByTestId('switch'));

		expect(onClick).not.toHaveBeenCalled();
		expect(onChange).not.toHaveBeenCalled();
	});

	it('should change active color when using checkedColor prop', () => {
		const { getByTestId } = render(
			<Switch data-testid="switch" defaultChecked checkedColor="black" />,
		);

		expect(getByTestId('switch').style.backgroundColor).toBe('black');
	});

	it('should change inactive active color when using unCheckedColor prop', () => {
		const { getByTestId } = render(
			<Switch data-testid="switch" unCheckedColor="white" />,
		);

		expect(getByTestId('switch').style.backgroundColor).toBe('white');
	});

	it('should apply active color to loading icon', () => {
		const { container } = render(
			<Switch
				data-testid="switch"
				loading
				defaultChecked
				checkedColor="red"
			/>,
		);

		expect(container).toMatchSnapshot();
	});

	it('should apply inactive color to loading icon', () => {
		const { container } = render(
			<Switch data-testid="switch" loading={false} checkedColor="red" />,
		);

		expect(container).toMatchSnapshot();
	});

	it('should change size when using size prop', () => {
		const { getByTestId } = render(
			<Switch data-testid="switch" size={20} />,
		);

		expect(getByTestId('switch').style.fontSize).toEqual('20px');
	});
});
