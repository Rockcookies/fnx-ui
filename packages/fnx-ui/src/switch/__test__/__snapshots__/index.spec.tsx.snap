// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Switch/> should apply active color to loading icon 1`] = `
<div>
  <div
    aria-checked="true"
    class="fnx-switch fnx-switch--on fnx-switch--loading"
    data-testid="switch"
    role="switch"
    style="background-color: red;"
  >
    <span
      class="fnx-switch__handle"
    >
      <span
        class="fnx-icon fnx-switch__loading"
        style="color: red;"
      >
        <svg
          aria-hidden="true"
          class="fnx-icon__spinner"
          fill="currentColor"
          focusable="false"
          height="1em"
          viewBox="25 25 50 50"
          width="1em"
        >
          <circle
            cx="50"
            cy="50"
            fill="none"
            r="20"
          />
        </svg>
      </span>
    </span>
    <span
      class="fnx-switch__inner"
    >
      <span
        class="fnx-switch__children"
      />
    </span>
  </div>
</div>
`;

exports[`<Switch/> should apply inactive color to loading icon 1`] = `
<div>
  <div
    aria-checked="false"
    class="fnx-switch"
    data-testid="switch"
    role="switch"
  >
    <span
      class="fnx-switch__handle"
    />
    <span
      class="fnx-switch__inner"
    >
      <span
        class="fnx-switch__children"
      />
    </span>
  </div>
</div>
`;
